File: master/linked_list_cycle.cpp
#include "header.h"

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

ListNode* create_list(int arr[], int n, int s){
    ListNode* result = NULL;
    ListNode* curr = NULL;
    ListNode* end = NULL;
    ListNode* start = NULL;
    for(int i=0; i<n; i++){
        ListNode* x = new ListNode(arr[i]);
        if(arr[i] == s) start=x;
        if(!result)
            result = curr = x;
        else{
            curr->next = x;
            curr = curr->next;
        }
        if(i==n-1) end=curr;
    }
    end->next = start;
    return result;
}

bool hasCycle(ListNode *head) {
    if(!head) return false;
    ListNode* slow = head;
    ListNode* fast = head;
    while(fast){
        slow = slow->next;
        fast = fast->next;
        if(!fast) return false;
        fast = fast->next;
        if(slow == fast) return true;
    }
    return false;
}

int main(){
    int arr[] = {0, 1, 2, 3};
    ListNode* l = create_list(arr, sizeof(arr)/sizeof(arr[0]), 0);
    bool r = hasCycle(l);
    printf("%s\n", r?"true":"false");
    return 0;
}

File: master/linked_list_cycle.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @return a boolean
    def hasCycle(self, head):
        slow = fast = head
        if head is None or head.next is None:
            return False
        while True:
            slow = slow.next
            fast = fast.next
            if not fast:
                return False
            fast = fast.next
            if not fast:
                return False
            if slow == fast:
                return True


if __name__ == "__main__":
    n1 = ListNode(1)    
    n2 = ListNode(2)    
    n3 = ListNode(3)    
    n4 = ListNode(4)    
    n5 = ListNode(5)    
    n1.next = n2
    n2.next = n3
    n3.next = n4
    n4.next = n2
    print Solution().hasCycle(n1)
